File: code/ex355.java
class Twitter {
    public int time;
    public Map<Integer, List<Integer>> tweets;
    public Map<Integer, Integer> id2time;
    public Map<Integer, Set<Integer>> follow;

    /** Initialize your data structure here. */
    public Twitter() {
        time = 0;
        tweets = new HashMap<Integer, List<Integer>>();
        id2time = new HashMap<Integer, Integer>();
        follow = new HashMap<Integer, Set<Integer>>();
    }
    
    /** Compose a new tweet. */
    public void postTweet(int userId, int tweetId) {
        id2time.put(tweetId, time++);
        
        List<Integer> list = tweets.getOrDefault(userId, new ArrayList<Integer>());
        list.add(tweetId);
        tweets.put(userId, list);
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    public List<Integer> getNewsFeed(int userId) {
        List<Integer> feed = new ArrayList<>();
        // Add his personal tweets to the feed
        if (tweets.containsKey(userId)) {
            for (int tweet : tweets.get(userId))
                feed.add(tweet);
        }
        
        // Add the tweets from people he follows
        if (follow.containsKey(userId)) {
            System.out.println(follow.get(userId));
            for (int f : follow.get(userId)) {
                if (tweets.containsKey(f)) {
                    for (int tweet : tweets.get(f)) {
                        feed.add(tweet);
                    }
                } 
            }
        }
        
        Collections.sort(feed, new Comparator<Integer>() {
            public int compare(Integer a, Integer b) {
                return id2time.get(b) - id2time.get(a);
            }
        });
        
        return feed.subList(0, Math.min(feed.size(), 10));
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    public void follow(int followerId, int followeeId) {
        if (followerId == followeeId)    return;
        Set<Integer> followees = follow.getOrDefault(followerId, new HashSet<Integer>());
        followees.add(followeeId);
        follow.put(followerId, followees);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    public void unfollow(int followerId, int followeeId) {
        if (!follow.containsKey(followerId) || followerId == followeeId)    return;
        Set<Integer> followees = follow.get(followerId);
        if (followees.contains(followeeId)) {
            followees.remove(followeeId);
        }
        follow.put(followerId, followees);
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */

File: code/ex355.cpp
class Twitter {
    unordered_map<int, vector<int>> user2tweets;
    unordered_map<int, int> id2time;
    unordered_map<int, unordered_set<int>> followees;
    int time = 0;
    
public:
    /** Initialize your data structure here. */
    Twitter() {
        
    }
    
    /** Compose a new tweet. */
    void postTweet(int userId, int tweetId) {
        user2tweets[userId].push_back(tweetId);
        id2time[tweetId] = time++;
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    vector<int> getNewsFeed(int userId) {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;    // {time, tweetId}
        
        for (int id : user2tweets[userId]) {    // Add tweets from the user
            pq.push({id2time[id], id});
            if (pq.size() > 10)    pq.pop();
        }
        
        for (int f : followees[userId]) {    // Add tweets from the user's followees;
            for (int id : user2tweets[f]) {
                pq.push({id2time[id], id});
                if (pq.size() > 10)    pq.pop();
            }
        }
        
        vector<int> feed;
        while (!pq.empty()) {
            feed.push_back(pq.top().second);
            pq.pop();
        }
        reverse(feed.begin(), feed.end());
        
        return feed;
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    void follow(int followerId, int followeeId) {
        if (followerId == followeeId)    return;
        followees[followerId].insert(followeeId);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    void unfollow(int followerId, int followeeId) {
        if (followees.count(followerId) == 0 || followees[followerId].count(followeeId) == 0)    return;
        followees[followerId].erase(followeeId);
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */